core/src/main/java/org/apache/flink/streaming/siddhi/SiddhiStream.java

```diff
@@ -124,19 +124,30 @@ public ExecutionSiddhiStream cql(String executionPlan) {
          * @return ExecutionSiddhiStream context
          */
         public ExecutionSiddhiStream cql(DataStream<ControlEvent> controlStream) {
-            DataStream<Tuple2<StreamRoute, Object>> unionStream = controlStream
-                .map(new NamedControlStream(ControlEvent.DEFAULT_INTERNAL_CONTROL_STREAM))
-                .broadcast()
-                .union(this.toDataStream())
+
+            // first of all we gather control events to the broadcasted stream
+            DataStream<Tuple2<StreamRoute, Object>> broadcastedControlStream = controlStream
+                    .map(new NamedControlStream(ControlEvent.DEFAULT_INTERNAL_CONTROL_STREAM))
+                    .broadcast();
+
+            // Control events are cut off on the AddRouteOperator stage
+            DataStream<Tuple2<StreamRoute, Object>> dataWithControlStream = this.toDataStream()
```

Comment:
I like your way of solving this. I do have also the logic for solving this like - 
```
public ExecutionSiddhiStream cql(DataStream<ControlEvent> controlStream) {
   DataStream<Tuple2<StreamRoute, Object>> unionStream = controlStream
                .map(new NamedControlStream(ControlEvent.DEFAULT_INTERNAL_CONTROL_STREAM))
                .broadcast()
                .union(this.toDataStream())
                .transform("add route transform",
                    SiddhiTypeFactory.getStreamTupleTypeInformation(TypeInformation.of(Object.class)),
                    new AddRouteOperator(getCepEnvironment().getDataStreamSchemas()));
DataStream<Tuple2<StreamRoute,Object>> controlEventStream = unionStream.filter(x ->x.f0.isBroadCastPartitioning()) 
DataStream<Tuple2<StreamRoute,Object>> eventStream = unionStream.filter(x -> !(x.f0.isBroadCastPartitioning()))

DataStream<Tuple2<StreamRoute, Object>> partitionedStream = new DataStream<>(
                    eventStream.getExecutionEnvironment(),
                    new PartitionTransformation<>(eventStream.getTransformation(),
                            new DynamicPartitioner()));

return new ExecutionSiddhiStream(partitionedStream.union(controlEventStream), null, getCepEnvironment());
```
However, I didn't change the AddRouteOperator logic. This thing will also work, I think this approach will be more overhead. Want your thought to it.

Reply:
I compared two solutions if they build the same JobGraphs or not.